sdk/servicebus/azure-servicebus/azure/servicebus/_common/utils.py

```diff
@@ -250,3 +250,44 @@ def shutdown(self, wait=True):
         :type wait: bool
         """
         self.executor.shutdown(wait=wait)
+
+
+class AutoComplete(object):
+    """
+    Provides a ContextManager to complete a message on successful exit from the scope,
+    calling `abandon()` otherwise, for instance if an exception is thrown.
+
+    :param message: The message to auto-complete or abandon.
+    :type message: ~azure.servicebus.ReceivedMessage
+
+    .. admonition:: Example:
+
+        .. literalinclude:: ../samples/sync_samples/sample_code_servicebus_async.py
+            :start-after: [START auto_complete_sync]
+            :end-before: [END auto_complete_sync]
+            :language: python
+            :dedent: 4
+            :caption: Automatically complete a message
+    """
+    def __init__(self, message):
+        self.message = message
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if traceback:
+            try:
+                if not self.message._settled:
+                    _log.info("Attempting to auto-abandon message due to exception: " + str(value))
+                    self.message.abandon()
+            except MessageAlreadySettled as e:
```

Comment:
We may want to catch other exceptions here... I would say in the case of exit-via error there's quite a high likelihood we'll be unable to settle. Though this is something we add in a subsequent preview if we feel it's necessary.

Reply:
I initially caught MessageSettleFailed as well but wanted to start as conservative as possible to avoid semantics that might be harder to back out.  Am totally open to reassessing as we see what rough edges emerge from consuming this more actively.